src/cnaas_nms/confpush/firmware.py

```diff
@@ -0,0 +1,289 @@
+
+    try:
+        res = task.run(netmiko_send_command, command_string='enable',
+                       expect_string='.*#')
+        print_result(res)
+
+        res = task.run(netmiko_send_command,
+                       command_string=firmware_download_cmd.replace("//", "/"),
+                       delay_factor=10,
+                       max_loops=3600)
+        print_result(res)
+    except Exception as e:
+        logger.info('{} failed to download firmware: {}'.format(
+            task.host.name, e))
+        raise e
+
+    return "Firmware download done."
+
+
+def arista_firmware_activate(task, filename: str) -> None:
+    """
+    NorNir task to modify the boot config for new firmwares.
+
+    Args:
+        task: NorNir task
+        filename: Name of the new firmware image
+
+    Returns:
+        Nope.
+
+    """
+    try:
+        boot_file_cmd = 'boot system flash:{}'.format(filename)
+
+        res = task.run(netmiko_send_command, command_string='enable',
+                       expect_string='.*#')
+        print_result(res)
+
+        res = task.run(netmiko_send_command, command_string='conf t',
+                       expect_string='.*config.*#')
+        print_result(res)
+
+        res = task.run(netmiko_send_command, command_string=boot_file_cmd)
+        print_result(res)
+
+        res = task.run(netmiko_send_command, command_string='end',
+                       expect_string='.*#')
+        print_result(res)
+
+        res = task.run(netmiko_send_command,
+                       command_string='show boot-config | grep -o "\\w*{}\\w*"'.format(filename))
+        print_result(res)
+
+        if not isinstance(res, MultiResult):
+            raise Exception('Could not check boot-config on {}'.format(task.host.name))
+
+        if res.result != filename:
+            raise Exception('Firmware not activated properly on {}'.format(task.host.name))
+
+    except Exception as e:
+        logger.exception('Failed to activate firmware: {}'.format(str(e)))
+        raise e
+
+    return "Firmware activate done."
+
+
+def arista_device_reboot(task) -> None:
+    """
+    NorNir task to reboot a single device.
+
+    Args:
+        task: NorNir task.
+
+    Returns:
+        Nothing.
+
+    """
+    try:
+        res = task.run(netmiko_send_command, command_string='enable',
+                       expect_string='.*#')
+        print_result(res)
+
+        res = task.run(netmiko_send_command, command_string='write',
+                       expect_string='.*#')
+        print_result(res)
+
+        res = task.run(netmiko_send_command, command_string='reload force',
+                       max_loops=2,
+                       expect_string='.*')
+    except Exception as e:
+        logger.exception('Failed to reboot switch {}: {}'.format(task.host.name,
+                                                            str(e)))
+        raise e
+
+    return "Device reboot done."
```

Comment:
Is the reboot done here as in the device has come back online again already, or has the reboot just been started?

Reply:
Done as in we've issued the reload command.